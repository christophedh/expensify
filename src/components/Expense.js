import React from 'react'
import { connect } from 'react-redux'
import { startRemoveExpense } from '../actions/expense'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'
export const Expense = ({ dispatch, id, description, amount, createdAt }) => (
    <div className="list-wrapper">
        <Link to={`/edit/${id}`} className="list-item">
            <div>
                <h3 className="list-item__title">{description}</h3>
                <span className="list-item__sub-title">
                    {moment(createdAt).format('LL')}
                </span>
            </div>
            <h3 className="list-item__data">
                <span className="show-for-desktop">Amount</span>
                {numeral(amount).format('$0,0.00')} <br />
            </h3>
        </Link>
        <button
            className="button button--secondary button--list-item"
            onClick={() => {
                dispatch(startRemoveExpense(id))
            }}
        >
            Remove
        </button>
    </div>
)

export default connect()(Expense)
