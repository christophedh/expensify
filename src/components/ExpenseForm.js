import React from 'react'
import moment from 'moment'
import { addExpense, editExpense } from '../actions/expense'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import { connect } from 'react-redux'
const uuidv4 = require('uuid/v4')

class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false
    }
    componentDidMount() {
        const { expense } = this.props

        if (expense) {
            this.setState(() => ({
                ...expense,
                createdAt: moment(expense.createdAt)
            }))
        }
    }
    onDescriptionChange = e => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = e => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = e => {
        const amount = e.target.value

        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = createdAt => {
        this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = e => {
        e.preventDefault()
        const { dispatch, expense } = this.props
        const { description, note, amount, createdAt } = this.state
        if (expense) {
            dispatch(
                editExpense({
                    expenseId: expense.id,
                    expenseUpdate: {
                        description,
                        note,
                        amount: parseFloat(amount),
                        createdAt: createdAt.unix()
                    }
                })
            )
        } else {
            dispatch(
                addExpense({
                    id: uuidv4(),
                    description,
                    note,
                    amount: parseFloat(amount),
                    createdAt: createdAt.unix()
                })
            )
        }
        this.props.onSubmitCompleted()
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        id={uuidv4()}
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button>
                        {this.props.expense ? 'Edit Expense' : 'Add Expense'}
                    </button>
                </form>
            </div>
        )
    }
}

export default connect()(ExpenseForm)
