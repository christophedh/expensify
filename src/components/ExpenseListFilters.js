import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filter'
import { DateRangePicker } from 'react-dates'
import { setStartDate, setEndDate } from '../actions/filter'
import moment from 'moment'

export class ExpenseListFilters extends React.Component {
    state = {
        startDate: undefined,
        endDate: undefined,
        focusedInput: undefined
    }
    componentDidMount() {
        const {
            filters: { startDate, endDate }
        } = this.props

        this.setState({
            startDate: moment(startDate),
            endDate: moment(endDate)
        })
    }

    onTextChange = e => {
        const text = e.target.value
        this.props.setTextFilter(text)
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.setState({
            startDate,
            endDate
        })
        this.props.setDates(startDate.valueOf(), endDate.valueOf())
    }

    setSortBy = e => {
        const sortBy = e.target.value

        if (sortBy === 'date') {
            this.props.sortByDate()
        } else if (sortBy === 'amount') {
            this.props.sortByAmount()
        }
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            placeholder="Search expense"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.setSortBy}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.state.startDate}
                            startDateId={'f2ea0227-995f-4846-a900-24dbbad9f022'}
                            endDate={this.state.endDate}
                            endDateId={'2ea0227-995f-4846-a900-24dbbad9f023'}
                            isOutsideRange={() => false}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.focusedInput}
                            onFocusChange={focusedInput =>
                                this.setState({ focusedInput })
                            }
                        />
                    </div>
                </div>
                <br />
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    setDates: (startDate = 0, endDate = 0) => {
        dispatch(setStartDate(startDate))
        dispatch(setEndDate(endDate))
    },
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
