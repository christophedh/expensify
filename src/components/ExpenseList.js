import React from 'react';
import { connect } from 'react-redux';
import Expense from './Expense';
import selectExpenses from '../selectors/expensesSelector';

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map((expense) => {
      return <Expense key={expense.id} {...expense} />;
    })}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
