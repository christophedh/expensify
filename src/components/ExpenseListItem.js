import React from 'react'
import { connect } from 'react-redux'
import Expense from './Expense'
import selectExpenses from '../selectors/expensesSelector'
export const ExpenseList = props => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expense</div>
            <div className="show-for-mobile">Amount</div>
            <div className="show-for-desktop">Expenses</div>
        </div>
        <div className="list-body">
            {props.expenses.length ? (
                props.expenses.map(expense => {
                    return <Expense key={expense.id} {...expense} />
                })
            ) : (
                <div className="list-item list-item__message">No expenses</div>
            )}
        </div>
    </div>
)

const mapStateToProps = state => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)
