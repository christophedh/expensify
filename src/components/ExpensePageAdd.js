import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
const AddExpensePage = props => {
    const {
        match: {
            params: { id }
        },
        history
    } = props
    console.log(id)
    return (
        <div>
            <h1>Add an expense</h1>
            <ExpenseForm
                onSubmitCompleted={() => {
                    history.push('/')
                }}
            />
        </div>
    )
}

export default connect()(AddExpensePage)
