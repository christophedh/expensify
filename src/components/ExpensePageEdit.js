import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expense'

export class EditExpensePage extends React.Component {
    onSubmit = expenseUpdate => {
        this.props.editExpense(expenseUpdate)
        this.props.history.push('/')
    }
    onDelete = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1>Expense form</h1>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    onDelete={this.onDelete}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: expenseUpdate => dispatch(editExpense(expenseUpdate)),
    removeExpense: id => dispatch(startRemoveExpense(id))
})

const mapStateToProps = ({ expenses }, props) => ({
    expense: expenses.find(expense => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
