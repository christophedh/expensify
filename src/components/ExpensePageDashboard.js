import React from 'react'
import ExpenseListItem from './ExpenseListItem'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary'

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseListItem />
    </div>
)

export default ExpenseDashboardPage
