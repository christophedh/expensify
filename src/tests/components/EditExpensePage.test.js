import React from 'react'
import { shallow } from 'enzyme'
import ExpenseForm from '../../components/ExpenseForm'
import { EditExpensePage } from '../../components/ExpensePageEdit'
import expenses from '../fixtures/expensesFixtures'
let wrapper, editExpense, history, removeExpense

beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[0]}
            editExpense={editExpense}
            removeExpense={removeExpense}
            history={history}
        />
    )
})

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
    const { id, expenseUpdate } = expenses[0]
    const data = { expenseId: id, expenseUpdate }
    wrapper.find(ExpenseForm).prop('onSubmit')(data)
    expect(editExpense).toHaveBeenCalledWith(data)
    expect(history.push).toHaveBeenCalledWith('/dashboard')
})

test('should handle removeExpense', () => {
    wrapper.find(ExpenseForm).prop('onDelete')()
    expect(removeExpense).toHaveBeenCalledWith(expenses[0].id)
    expect(history.push).toHaveBeenCalledWith('/dashboard')
})
