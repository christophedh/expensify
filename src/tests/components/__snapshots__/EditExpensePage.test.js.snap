// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 123,
          "createdAt": 0,
          "description": "gum",
          "id": "1",
          "note": "gum",
        }
      }
      onDelete={[Function]}
      onSubmit={[Function]}
    />
  </div>
</div>
`;
